For check linked list

Gaydarenko opened the pull request:
Я еще не доделал. Но у меня получилось сдвинуться с мертвой точки (было критичное непонимание).

Gaydarenko committed:
Release def append and correct def __str__

Gaydarenko committed:
Изменил идею вставки:
def insert - проверяет значение индекса и производит вставку только индекс в диапазоне длины списка. Если же индекс больше длины списка или список пуст, то производит вызов метода def append

Gaydarenko committed:
in def insert add insert with index == 0

Gaydarenko committed:
ничего интересного - нужен переместиться на другой комп

Gaydarenko committed:
release def insert - insert in center

Gaydarenko committed:
release def find

Gaydarenko committed:
add comment about different solutions

EnotYoyo commented on linkedlist.py:
```diff
@@ -78,6 +97,16 @@ def __iter__(self):
             yield current_node.value
             current_node = current_node.next
 
+    # def __reversed__(self):
```
А что не так с методом?

Gaydarenko replied:
Пока не понял что этот метод должен делать. Начал делать под одно действие, потом увидел, что возможно нужно совершенно другое. 
И переключился пока на другие методы

EnotYoyo commented on linkedlist.py:
```diff
@@ -88,7 +117,19 @@ def clear(self) -> None:
+        :param node: node value
+        :return: node index or -1 if the value is not found
+        """
+        index = -1
```
Используйтке enumerate, так код будет чище :)

Gaydarenko replied:
fixed

EnotYoyo commented:
> было критичное непонимание

А что именно было непонятно? 

Пока что все очень хорошо. Единственное - можете сделать дополнительный метод, который будет возвращать итератор по нодам (аналогия __iter__ только которые возвращает сами ноды, а не данные). Это позволит не писать for через длину списка.

Gaydarenko committed:
есть пара проблем

Gaydarenko committed:
made the code cleaner in def find

Gaydarenko committed:
release "def remove"

Gaydarenko committed:
in Node use weak references for prev

Gaydarenko committed:
upgrade:
1) "def remove" in linkedlist.py
2) property "def next" in node.py
3) property "def prev" in node.py
release:
1) "def delete"

Gaydarenko committed:
uncommented "def __reversed__"

Gaydarenko committed:
"def verification_index" decorated by staticmethod

EnotYoyo commented on linkedlist.py:
```diff
@@ -87,21 +116,104 @@ def clear(self) -> None:
+            self.__head = self.__head.next
+
+        elif index == self.__len - 1:
+            self.__tail = self.__tail.prev
```
А у нового tail все еще останется ссылка next

Gaydarenko replied:
fix

EnotYoyo commented on linkedlist.py:
```diff
@@ -87,21 +116,104 @@ def clear(self) -> None:
+            raise IndexError
+
+        if index == 0:
+            self.__head = self.__head.next
```
А занулить у нового head prev? Тут спасает то, что prev это weakref, но лучше явно занулить

Gaydarenko replied:
Изначально так и задумывалось - понятный и привычный путь. Но как раз по вашему совету попробовал в коде использовать weakref. Переработал и другие части, чтобы убрать избыточность.  И это было моей небольшой гордостью, что понял и сумел применить.
Хочется понять эту грань где избыточность. Почему лучше явно занулить?  Это для удобства чтения кода или возможности других багов, про которые я пока не знаю?

EnotYoyo replied:
Это здорово, что Вы здесь сделали это осознано. Занулять стоит, поскольку это корректно с точки зрения логики. Конкретно этот класс не управляет сборщиком мусора, Вы не можете быть уверен, что пользователь Вашего класса [не выключил](https://docs.python.org/3/library/gc.html#gc.disable) сборщик мусора в своей программе.

EnotYoyo commented on linkedlist.py:
```diff
@@ -87,21 +116,104 @@ def clear(self) -> None:
 
-    def find(self, node):
-        ...
+    def find(self, node: Node) -> int:
```
зачем тут нужен именно Node? Ищите сразу по данным.

Gaydarenko replied:
В некоторых функциях указана переменная на входе "value", а у некоторых "node". Предположил, что речь идет о подаче на вход именно ноды. К тому же было интересно попробовать использовать ноду (чтобы перестать их опасаться).

EnotYoyo commented on linkedlist.py:
```diff
@@ -87,21 +116,104 @@ def clear(self) -> None:
+        else:
+            return -1
+
+    def remove(self, node: Node):
```
Аналогично, ищите по данным. И у Вас описание из документации не соответствует тому, что делает функция (докстринг говорит, что удаляется первая нода с такими данными, а на самом деле это не так, поскольку вы проверяете отдельно tail)

Gaydarenko replied:
fix

Gaydarenko committed:
fix "def delete" - new tail haven't link next

Gaydarenko committed:
fix "def remove" - Remove the FIRST occurrence

Gaydarenko committed:
add tests for Linked List

EnotYoyo commented:
Поправьте зануление и зачтено :+1:

Gaydarenko committed:
fix "def delete" - добавил принудительное зануление self.head
add docstring to "def delete"
